rpc: remove DRPC dependency from `Peer` and `Connection` generics

cthumuluru-crdb opened the pull request:
Although `Peer` and `Connection` are generic and intended to support both gRPC and DRPC connections, the current implementation has a hardcoded dependency on DRPC. As a result, a DRPC connection is always dialed, regardless of the intended type. This PR removes the direct dependency on DRPC, allowing the appropriate connection type (gRPC or DRPC) to be used based on the generic parameter.

Epic: CRDB-48923
Fixes: none
Release note: none

cthumuluru-crdb committed:
rpc: remove DRPC dependency from Peer and Connection generics

Although `Peer` and `Connection` are generic and intended to support both
gRPC and DRPC connections, the current implementation has a hardcoded
dependency on DRPC. As a result, a DRPC connection is always dialed,
regardless of the intended type. This PR removes the direct dependency
on DRPC, allowing the appropriate connection type (gRPC or DRPC) to be
used based on the generic parameter.

Epic: CRDB-48923
Fixes: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/148609)
<!-- Reviewable:end -->

cthumuluru-crdb committed:
kv,server: extract RangeFeed and register with DRPC server

This is a follow-up to #145195, where we extracted KVBatch from the
Internal service. Here, we do the same with RangeFeed. The TL;DR is
that smaller services are easier to maintain and can be more smoothly
migrated to DRPC.

We also enable this service on the DRPC server. This is controlled by
`rpc.experimental_drpc.enabled` (off by default).

This change is part of a series and is similar to: #146926

Note: This only registers the service; the client is not updated to use the
DRPC client, so this service will not have any functional effect.

Epic: CRDB-48925
Release note: None

shubhamdhama commented on pkg/rpc/nodedialer/nodedialer.go:
nit: we can use `Dial` and `DialDRPC`  IMO

shubhamdhama commented on pkg/rpc/stream_pool.go:
```diff
@@ -344,8 +344,3 @@ type BatchStreamClient = streamClient[*kvpb.BatchRequest, *kvpb.BatchResponse]
 type DRPCBatchStreamPool = streamPool[*kvpb.BatchRequest, *kvpb.BatchResponse, drpc.Conn]
 
 type DRPCBatchStreamClient = streamClient[*kvpb.BatchRequest, *kvpb.BatchResponse]
```
We can remove this line now.

shubhamdhama reviewed (commented):
LGTM, left some comments. I'll take another look and approve it.

tbg commented on pkg/rpc/nodedialer/nodedialer.go:
```diff
@@ -220,71 +196,37 @@ func (n *Dialer) DialInternalClient(
-		}
-		err = errors.Wrapf(err, "failed to connect to n%d at %v", nodeID, addr)
-		return nil, nil, err
+		return nilConn, nil, errors.Wrapf(err, "failed to connect to n%d at %v", nodeID, addr)
```
Not necessarily for this PR, but it would be good if we could tell whether drpc or gRPC were used in the error message.

cthumuluru-crdb commented:
bors r+

cthumuluru-crdb commented:
bors r+
